File: 3rd/src/ch5_03/Hoge.java
package ch5_03;

import java.time.LocalDate;
import java.time.temporal.TemporalAdjuster;
import java.util.function.Predicate;

/**
 * Q：Predicate<LocalDate>を受け取りTemporalAdjusterを返すnextメソッドを実装しなさい。
 *
 * @see http://docs.oracle.com/javase/jp/8/api/java/time/temporal/
 *      TemporalAdjuster.html
 */
public class Hoge {

	/**
	 * 指定された条件式を満たす「今日の日付の後」の日付。「今日の日付の後」はこのメソッド内で実現すること。
	 *
	 * @param predicate
	 *            条件式
	 * @return 条件に合致したTemporalAdjusterのオブジェクト
	 * @throws java.lang.IllegalArgumentException
	 *             本日より1000日以内に該当の日にちが存在しない
	 */
	public static TemporalAdjuster next(Predicate<LocalDate> predicate) {

		return t -> {
			LocalDate w = (LocalDate) t;
			int max = 1000; // 1000日以内
			int i = 0;

			for (i = 0; i < max; i++) {
				w = w.plusDays(1); // 今日より後にあって・・・の実装。
				if (predicate.test(w)) {
					break;
				}
			}

			if (i >= max) {
				throw new IllegalArgumentException("本日より" + max + "日以内に該当の日にちは存在しません。");
			}
			// LocalDateはTemporalAdjusterインタフェースの実装である。
			return w;
		};
	}

	public static void main(String[] args) {

		LocalDate today = LocalDate.now();

		// http://docs.oracle.com/javase/jp/8/api/java/time/LocalDate.html#with-java.time.temporal.TemporalAdjuster-
		LocalDate weekday = today.with(Hoge.next(w -> w.getDayOfWeek().getValue() < 6));
		System.out.println(weekday);
	}

}

File: 4th/src/ch7_03/hoge.js
#!/usr/bin/jjs
/**
 * Q：次の呼び出しを行いなさい。
 * A：以下のアウトプットが得られた。
 *
jjs> var b = new java.math.BigInteger('12345678790987654321')
jjs> b
12345678790987655000

 *
 * Q：b.mod(java.math.BigInteger.TEN)の値は何ですか。
 * A：以下のアウトプットが得られた。
 *
jjs> b.mod(java.math.BigInteger.TEN)
1
 *
 * Q：bはなぜ奇妙に表示されるのですか。
 * A：
 *
 */
var b = new java.math.BigInteger('12345678790987654321')
print(b)
print(b.mod(java.math.BigInteger.TEN))

/**
 *
12345678790987655000
1
 *
 *
 *
 * BigInteger(String val)
 * BigIntegerの10進String表現をBigIntegerに変換します。
 * @see http://docs.oracle.com/javase/jp/8/api/java/math/BigInteger.html#BigInteger-java.lang.String-
 */
